src/player.c

```diff
@@ -496,3 +531,111 @@ void *BarPlayerThread (void *data) {
 	return (void *) pret;
 }
 
+void *BarAoPlayThread (void *data) {
+	assert (data != NULL);
+
+	player_t *const player = data;
+
+	AVFrame *frm = NULL;
+	AVPacket *pkt = NULL;
+
+	bool quit = false;
+
+	while (!shouldQuit (player) && !quit) {
+		struct _QUEUE_ELEMT * queue_element = NULL;
+		while (!shouldQuit (player) && !quit && queue_element == NULL) {
+
+			/* force pausing until the queue size get 256 or decoding is done*/
+			pthread_mutex_lock (&player->aoplay_lock);
+			if (player->forcePause && !player->queueDone) {
+				do {
+					pthread_cond_wait (&player->aoplay_cond, &player->aoplay_lock);
+				} while (player->forcePause && !player->queueDone);
+			}
+			pthread_mutex_unlock (&player->aoplay_lock);
+
+			/* pausing by user*/
+			pthread_mutex_lock (&player->lock);
+			if (player->doPause) {
+				av_read_pause (player->fctx);
```

Comment:
You can’t call ffmpeg functions here, since fctx is not protected by a mutex in the decoder.